### PR TITLE
#3842 Introduce Ohara's partitioner

### PR DESCRIPTION
**Pull request by chia7712**
### Checklist
- [x] [Review Contribution Guidelines](https://ohara.readthedocs.io/en/latest/contributing.html)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #3842)
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-kafka/src/main/java/com/island/ohara/kafka/RowDefaultPartitioner.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+package com.island.ohara.kafka;
+
+import com.island.ohara.common.data.Row;
+import com.island.ohara.common.setting.TopicKey;
+
+public class RowDefaultPartitioner extends RowPartitioner {
+
+  @Override
+  public int partition(TopicKey topicKey, Row row, Cluster cluster) {
+    throw new UnsupportedOperationException("RowDefaultPartitioner does NOT support partition");
```
Comment (jackyoh) [inline ohara-kafka/src/main/java/com/island/ohara/kafka/RowDefaultPartitioner.java]: 這個方法未來會實作 OHARA 的 partition 邏輯嗎？

**Comment by chia7712** [comment]
retry collieIT

File: ohara-kafka/src/main/java/com/island/ohara/kafka/RowDefaultPartitioner.java (hunk 1 above)
Comment (chia7712) [reply ohara-kafka/src/main/java/com/island/ohara/kafka/RowDefaultPartitioner.java]: 不會 因為default partition仰賴super class去提供基本的實作，這邊有一點尷尬的地方是要避免不必要的轉換 (bytes -> row)

我們有暴露一個package-private的方法 - byDefault -> 用來說只想要預設的partition邏輯，這樣就不會進行bytes的傳換

Comment (jackyoh) [reply ohara-kafka/src/main/java/com/island/ohara/kafka/RowDefaultPartitioner.java]: 所以 default partition 的程式邏輯應該會在 RowPartitioner.java 的 partition 裡

Comment (chia7712) [reply ohara-kafka/src/main/java/com/island/ohara/kafka/RowDefaultPartitioner.java]: 沒錯，讓我加一些註解好了

File: ohara-kafka/src/main/java/com/island/ohara/kafka/RowPartitioner.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+/**
+ * The partition distribution is up to RowPartitioner. Implementing custom RowPartitioner is able to
+ * define the distribution of data and you can supply the custom balancer strategy.
+ *
+ * <p>By default, the distribution is calculated by hash.
+ */
+public abstract class RowPartitioner implements Partitioner {
+
+  private final Partitioner defaultPartitioner = new DefaultPartitioner();
```
Comment (vitojeng) [inline ohara-kafka/src/main/java/com/island/ohara/kafka/RowPartitioner.java]: Due to we have another class: `RowDefaultPartitioner`
`defaultPartitioner` rename to `kafkaDefaultPartitioner` could be avoid confuse IMHO.

Comment (chia7712) [reply ohara-kafka/src/main/java/com/island/ohara/kafka/RowPartitioner.java]: good point :)

**Comment by chia7712** [comment]
retry configurator

**Pull request merged**